doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+ in
+ let from_ext = function
+   | ExtTerminal s -> Sterm s
+   | ExtNonTerminal (s, _) -> to_doc_sym s
+ in
+ List.map from_ext prod
+
+let rec cvt_gram_sym = function
+  | GSymbString s -> Sterm s
+  | GSymbQualid (s, level) ->
+    Snterm (match level with
+           | Some str -> s ^ str
+           | None -> s)
+  | GSymbParen l -> Sparen (cvt_gram_sym_list l)
+  | GSymbProd ll ->
+    let cvt = List.map cvt_gram_prod ll in
+    (match cvt with
+    | (Snterm x :: []) :: [] -> Snterm x
+    | (Sterm x :: []) :: []  -> Sterm x
+    | _ -> Sprod cvt)
+
+and cvt_gram_sym_list l =
+  let get_sym = function
+    | GSymbQualid (s, level) -> s
+    | _ -> ""
+  in
+
+  match l with
+  | GSymbQualid ("LIST0", _) :: s :: GSymbQualid ("SEP", _) :: sep :: tl ->
+    Slist0sep (cvt_gram_sym s, cvt_gram_sym sep) :: cvt_gram_sym_list tl
+  | GSymbQualid ("LIST1", _) :: s :: GSymbQualid ("SEP", _) :: sep :: tl ->
+    Slist1sep (cvt_gram_sym s, cvt_gram_sym sep) :: cvt_gram_sym_list tl
+  | GSymbQualid ("LIST0", _) :: s :: tl ->
+    Slist0 (cvt_gram_sym s) :: cvt_gram_sym_list tl
+  | GSymbQualid ("LIST1", _) :: s :: tl ->
+    Slist1 (cvt_gram_sym s) :: cvt_gram_sym_list tl
+  | GSymbQualid ("OPT", _) :: s :: tl ->
+    Sopt (cvt_gram_sym s) :: cvt_gram_sym_list tl
+  | GSymbQualid ("IDENT", _) :: s2 :: tl when get_sym s2 = "" ->
+    cvt_gram_sym s2 :: cvt_gram_sym_list tl
+  | GSymbQualid ("ADD_OPT", _) :: tl ->
+    (Sedit "ADD_OPT") :: cvt_gram_sym_list tl
+  | GSymbQualid ("NOTE", _) :: GSymbQualid (s2, l) :: tl ->
+    (Sedit2 ("NOTE", s2)) :: cvt_gram_sym_list tl
+  | GSymbQualid ("USE_NT", _) :: GSymbQualid (s2, l) :: tl ->
+    (Sedit2 ("USE_NT", s2)) :: cvt_gram_sym_list tl
+  | hd :: tl ->
+    cvt_gram_sym hd :: cvt_gram_sym_list tl
+  | [] -> []
+
+and cvt_gram_prod p =
+ List.concat (List.map (fun x -> let _, gs = x in cvt_gram_sym_list gs)  p.gprod_symbs)
+
+
+let filename_tail s =
+  try
+    let i = String.rindex s '/' in
+    String.sub s (i + 1) ((String.length s) - (i + 1))
+  with Not_found -> s
+
+let add_symdef nt file symdef_map =
+  let ent =
+    try
+      StringMap.find nt !symdef_map
+    with Not_found -> []
+  in
+  symdef_map := StringMap.add nt (filename_tail file::ent) !symdef_map
+
+let rec edit_SELF nt_name cur_level next_level right_assoc prod =
+  let len = List.length prod in
+  List.mapi (fun i sym ->
+    match sym with
+    | Snterm s -> begin match s with
+      | s when s = nt_name || s = "SELF" ->
+        if i = 0 then Snterm next_level
+        else if i+1 < len then sym
+        else if right_assoc then Snterm cur_level else Snterm next_level
+      | "NEXT" -> Snterm next_level
+      | _ -> sym
+    end
+    | Slist1 sym -> Slist1 (List.hd (edit_SELF nt_name cur_level next_level right_assoc [sym]))
+    | Slist0 sym -> Slist0 (List.hd (edit_SELF nt_name cur_level next_level right_assoc [sym]))
+    | x -> x)
+  prod
+
+
+let mlg_files_ordered = [ (* in the order they are loaded by Coq *)
```

Comment:
This looks quite fragile, but I don't know how to do any better.

Reply:
Yeah, the best I could think of so far.  Maybe it will go away by relying exclusively on `orderedGrammar`.  But for now seems useful.